drud/ddev
# Configure Xdebug debugging sessions to be started by a trigger by default, fixes #3593, fixes #3731

## The Problem/Issue/Bug:

* #3593 
* #3731 

As an experienced PHP developer 
- I would like to be able to use all of Xdebug's features with zero configuration.
- I would like to use the development features without triggering a debugging session.
- I do not expect any errors or warnings when executing PHP scripts on the command line.
- I would like to be able to generate line and path coverage reports of my code using phpunit and xdebug.
- I expect to be able to start debugging session via the XDebug browser extension or from my IDE.

As a non experienced PHP developer 
- I would like to start a xdebug debugging session with minimal effort.
- I would like to use the improved error traces provided by xdebug.

## How this PR solves the problem:

This PR is based on the results of the issue #3731.

Xdebug is enabled by default with all of its features. A debugging session is no longer started automatically. 
This allows the usage of the Xdebug development features during development of your application (and not debugging it) as well as triggering alternative features like profiling or tracing.
In this mode you can enable debugging sessions by cookies (browser extension or bookmarks) or by environment variables on the cli (set by your IDE).
To enable the automatic start of a debugging session you can run `ddev xdebug autostart` or by setting the value `xdebug_autostart: true` in the .ddev/config.yml. 
This will start a debugging session on each request and on every started script on the command line. This is the current default behavior, when enabling xdebug.
The Xdebug extension can still be disabled by running `ddev xdebug off` or setting `xdebug: false` in the configuration.

### Architectural Decisions

#### Preferring "autostart" over "trigger" as a new mode
As discussed in #3731 the initial idea was to introduce a new mode "trigger" to provide new behavior to stay compatible to the current default behavior. 
I did some research and asked developers what they would expect would happen when they execute "ddev xdebug trigger", in most cases they could not imagine the concrete behavior or describe it precise. 
At the same time "ddev xdebug autostart" was easy to understand and describe for everyone.
The current default behavior of xdebug in ddev was also not expected by the asked developers (no ddev users).

#### Enabling Xdebug by default
Enabling Xdebug is a reasonable default as it provides useful development features like better var_dump function and traces for errors. 
Using Xdebug in a development environment is a common default. To test your application in a more production like environment you can still disable the extension.
With disabled autostart of debugging sessions there are no drawbacks in running Xdebug during development. 

#### Disabling autostart mode by default
Automatically starting debugging sessions is not always intended. 
There are serveral common situations in which you do not like to start a debugging session when Xdebug is active:
- Debugging your Application with xdebug extended var_dump and xdebug error traces (Of course NOBODY would do that)
- Generating line or path coverage reports on the command line
- Debugging a single request, controlled by your browser plugin instead of debuggin all requests
- Using the profiling ot tracing features of xdebug
- You like to debug requests in the browser but not on the command line
As debugging all requests on CLI and in browser seems to me a special use case this should not be the default.

#### Using phpenmod for switchting the configuration
phpenmod is a tool to enable or disable php modules. It is not meant for switching configurations. 
I decided to use it anyway, as it makes it easy to apply the configuration change for all php variants (Module, FPM and CLI) and all versions (5.6 - 8.1). 
Also the current xdebug configuration is already placed in a mods-available configuration, so using those files for configuration changes seems to be acceptable in the DDEV project.

#### Supporting Xdebug 2
As discussed in #3731 DDEV does not implement new features for xdebug 2.
However this pull request does support xdebug 2, because not supporting it would have added complexity to this pull request. 
Omitting support for xdebug 2 would cause errors executing `phpenmod xdebug_autostart` as the auto start configuration file would not exist for php version lower than 7.2. Also the behavior between xdebug 2 and xdebug 3 php versions would change, which would add complexity for the user and to the documentation. 

### Breaking Changes
Xdebug will be enabled by default. add `xdebug_enabled: false` to disable it.
Debugging session will not start automatically. You can trigger them by browser or environment variable on CLI, or you have to run `ddev xdebug autostart` to debug every request, this will equal the current behavior when xdebug is enabled.


## Manual Testing Instructions:

- Run `ddev config` in a new project
- When using an existing project, make sure there is no `xdebug_enabled: false` in `.ddev/config.yml`

### Default behavior
- Run `ddev xdebug status`
  - should return "xdebug enabled (debug,profile,trace)"
- Browse your website 
  - no debugging session should start
- Open your browser and start a debugging session by using a bookmark or the xdebug extension
  - a debugging session should start
- Run a debugging session on cli with your IDE
  - a debugging session should start
- Run any php script on cli
  - no debugging session should start and there should be no unexpected xdebug output

### Switching Mode
- Run `ddev xdebug autostart`
- Run `ddev xdebug status`
  - should return "xdebug enabled with autostart (debug,develop)"
- Open your application in browser
  - a debugging session should start
- Run any script on cli
  - a debugging session should start
- Run `ddev xdebug trigger`
  -  See default behavior

## Automated Testing Overview:
- xdebug_test.go
- ddevapp_test.go

## Related Issue Link(s):
- https://github.com/drud/ddev/issues/3731

## Release/Deployment notes:
N/A

## Open todos
- [x] Implement changes
- [x] Update Container
- [x] Update / add test caes

## Related

#### Xdebug bookmark generator
https://www.jetbrains.com/phpstorm/marklets/

#### Xdebug Helper Browser Extension
https://chrome.google.com/webstore/detail/xdebug-helper/eadndfjplgieldjbigjakmdgkmoaaaoc?hl=de

#### PHPStorm Documentation on debugging
https://www.jetbrains.com/help/phpstorm/zero-configuration-debugging.html#activate-debugger-on-server

#### Xdebug error traces
![image](https://user-images.githubusercontent.com/18308418/162589306-5765b1cb-1a6c-4157-984b-045b1cf3ae37.png)

#### Xdebug improved var_dumps
![image](https://user-images.githubusercontent.com/18308418/162589283-7c7b0ca4-b045-482d-b160-5aa22930fc5b.png)


<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3754"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
